dag_in_context/src/pretty_print.rs

```diff
@@ -0,0 +1,831 @@
+            }
+            // doesn't fold Tuple
+            Expr::Single(x) => {
+                let sub_expr = self.fold_expr(x, fold_when, to_rust);
+                Expr::Single(Rc::new(sub_expr))
+            }
+            Expr::Concat(x, y) => {
+                let left = self.fold_expr(x, fold_when, to_rust);
+                let right = self.fold_expr(y, fold_when, to_rust);
+                Expr::Concat(Rc::new(left), Rc::new(right))
+            }
+            Expr::Switch(x, inputs, _branches) => {
+                let cond = self.fold_expr(x, fold_when, to_rust);
+                let inputs = self.fold_expr(inputs, fold_when, to_rust);
+                let branches = _branches
+                    .iter()
+                    .map(|branch| Rc::new(self.fold_expr(branch, fold_when, to_rust)))
+                    .collect::<Vec<_>>();
+                let switch = Expr::Switch(Rc::new(cond), Rc::new(inputs), branches);
+                fold_or_plain(self, switch)
+            }
+            Expr::If(x, inputs, y, z) => {
+                let pred = self.fold_expr(x, fold_when, to_rust);
+                let inputs = self.fold_expr(inputs, fold_when, to_rust);
+                let left = self.fold_expr(y, fold_when, to_rust);
+                let right = self.fold_expr(z, fold_when, to_rust);
+                let if_expr = Expr::If(
+                    Rc::new(pred),
+                    Rc::new(inputs),
+                    Rc::new(left),
+                    Rc::new(right),
+                );
+                fold_or_plain(self, if_expr)
+            }
+            Expr::DoWhile(inputs, body) => {
+                let inputs = self.fold_expr(inputs, fold_when, to_rust);
+                let body = self.fold_expr(body, fold_when, to_rust);
+                let dowhile = Expr::DoWhile(Rc::new(inputs), Rc::new(body));
+                fold_or_plain(self, dowhile)
+            }
+            Expr::Arg(ty, assum) => {
+                let ty = handle_type(ty, self);
+                let assum_str = handle_assum(assum, self, fold_when, to_rust);
+                Expr::Arg(ty, Assumption::WildCard(assum_str))
+            }
+            Expr::Symbolic(_) => panic!("No symbolic should occur here"),
+        }
+    }
+}
+
+impl Expr {
+    pub fn pretty(&self) -> String {
+        let (term, termdag) = Rc::new(self.clone()).to_egglog();
+        let expr = termdag.term_to_expr(&term);
+        expr.to_sexp().pretty()
+    }
+
+    fn concat_helper(&self) -> Vec<String> {
+        match self {
+            Expr::Concat(lhs, rhs) => {
+                let mut lhs = lhs.as_ref().concat_helper();
+                let mut rhs = rhs.as_ref().concat_helper();
+                lhs.append(&mut rhs);
+                lhs
+            }
+            Expr::Single(expr) => {
+                let expr = Self::to_ast(expr.as_ref());
+                vec![expr]
+            }
+            _ => panic!("Not well formed Concat, expr not in Single"),
```

Comment:
This isn't correct- a concat could take, for example, a Loop as one of the expressions.

Reply:
Add a test for this case